File: cpp/15.三数之和.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>>res;
        if(nums.size()<3) return res;      
        sort(nums.begin(),nums.end());
        for(int i=0;i<nums.size()-2;++i)
        {
            if(i==0||(i>0&&nums[i]!=nums[i-1]))
            {
                int l=i+1,r=nums.size()-1;
                int sum=-nums[i];
                while(l<r)
                {
                    if(nums[l]+nums[r]==sum)
                    {
                        vector<int>temp;
                        temp.push_back(nums[i]);
                        temp.push_back(nums[l]);
                        temp.push_back(nums[r]);
                        res.push_back(temp);
                        while(l+1<r&&nums[l+1]==nums[l]) l++;
                        while(r-1>l&&nums[r-1]==nums[r]) r--;
                        l++;
                        r--;
                    }    
                    else if(nums[l]+nums[r]<sum)  l++;
                    else r--;
                }
            }
        }
        return res;
    }
};

File: javascript/15.三数之和.js
/*
 * @lc app=leetcode.cn id=15 lang=javascript
 *
 * [15] 三数之和
 */
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    var result = [];
    if(nums.length < 3) return result;

    nums.sort((a,b) => a - b);
    for(var i = 0; i < nums.length - 2; i++){
        if((i == 0)||(i > 0 && nums[i] != nums[i-1])){
            var start = i + 1, end = nums.length - 1;
            while(start < end){
                var temp =  nums[i] + nums[start] + nums[end];
                if(temp == 0){
                    var res = new Array(nums[i],nums[start],nums[end]);
                    result.push(res);
                    while(start + 1 < end && nums[start+1] == nums[start]) start++;
                    while(end - 1 > start && nums[end-1] == nums[end]) end--;
                    start++;
                    end--;
                }else if(temp < 0){
                    start++;
                }else{
                    end--;
                }
            }
        }  
    }
    return result;
};

